### ISSUE TITLE
Configuration singleton causes problem after class reloading

### ISSUE THREAD

**Issue by neworld** [Litote/kmongo#75]
For faster development, I am using hot class reloading instead of full JVM restart. Also, I am using tons of custom serializers to make structure friendlier to the MongoDB. The problem is, after each class reload, currently assigned mixins and modules doesn't work with newly loaded classes. So, I am adding new modules each time. Apparently, count of modules and mixins is increasing dramatically. 

However, I suppose there could be more cases it will not play nicely. My suggestion should be backward compatible:

```kotlin
inteface KMongoCodecConfig {
  val bsonMapper: ObjectMapper
  val extendedJsonMapper: ObjectMapper
  val bsonMapperCopy: ObjectMapper
}

object KMongoConfiguration: KMongoCodecConfig {
  //nothing changes
}

fun codecConfig(dsl: KMongoCodecConfigurator.() -> Unit): KMongoCodecConfig {
  // ....
}

object KMongo {
  fun createClient(addr: ServerAddress, options: MongoClientOptions, codecConfig: KMongoCodecConfig = KMongoConfiguration) {
    ....
  }
  //other methods should be adapted as well
}
```
Usage example:
```kotlin
val projectCodecConfig = codecConfig {
  addModule(VeryNeededModule())
}
KMongo.createClient(codecConfig = projectCodecConfig)
```

There are tons of other way to do that. Considering singleton is bad design in general, maybe would be acceptable to remove singleton in next major release?

**Comment by zigzago** [comment]
By the way there is a memory leak fixed in 3.8.3

**Comment by neworld** [comment]
Thanks for fixing this! I will test it soon
